File: json-parser/python_json_parser/src/parser.py
def parse_json(input_string):
    index = 0

    def consume_whitespace():
        nonlocal index
        while index < len(input_string) and input_string[index].isspace():
            index += 1

    def consume_char(char):
        nonlocal index
        consume_whitespace()
        if index < len(input_string) and input_string[index] == char:
            index += 1
            return True
        return False

    def parse_value():
        consume_whitespace()
        if input_string[index] == '"':
            return parse_string()
        elif input_string[index] == 't' or input_string[index] == 'f':
            return parse_boolean()
        elif input_string[index] == 'n':
            return parse_null()
        elif input_string[index].isdigit() or input_string[index] == '-':
            return parse_number()
        elif input_string[index] == '{':
            return parse_object()
        elif input_string[index] == '[':
            return parse_array()
        return False

    def parse_boolean():
        nonlocal index
        if input_string[index:index+4] == 'true':
            index += 4
            return True
        elif input_string[index:index+5] == 'false':
            index += 5
            return True
        return False

    def parse_null(): 
        nonlocal index 
        if input_string[index:index+4] == 'null':
            index += 4
            return True
        return False

    def parse_number():
        nonlocal index
        start = index 
        if input_string[index] == '-':
            index += 1
        while index < len(input_string) and input_string[index].isdigit():
            index +=1
        if index < len(input_string) and input_string[index] == '.':
            index += 1
            while index < len(input_string) and input_string[index].isdigit():
                index += 1
        return index > start

    def parse_string():
        nonlocal index
        if not consume_char('"'):
            return False
        while index < len(input_string) and input_string[index] != '"':
            if input_string[index] =='\\':
                index += 1
            index += 1
        if not consume_char('"'):
            return False
        return True

    def parse_key_value_pair():
        if not parse_string():
            return False
        if not consume_char(':'):
            return False
        if not parse_value():
            return False
        return True

    def parse_array():
        if not consume_char('['):
            return False
        consume_whitespace()
        if consume_char(']'):
            return True
        while True:
            if not parse_value():
                return False
            consume_whitespace()
            if not consume_char(','):
                break
        if not consume_char(']'):
            return False
        return True
            

    def parse_key_value_pair():
        if not parse_string():
            return False
        consume_whitespace()
        if not consume_char(':'):
            return False
        consume_whitespace()
        if not parse_value():
            return False
        return True

    def parse_object():
        if not consume_char('{'):
            return False
        consume_whitespace()
        if consume_char('}'):
            return True
        while True:
            if not parse_key_value_pair():
                return False
            consume_whitespace()
            if not consume_char(','):
                break
            consume_whitespace()
        if not consume_char('}'):
            return False
        return True

    return parse_object()

File: json-parser/ts_json_parser/src/parser.ts
export function parseJson(input: string): boolean {
    let index = 0;

    function consumeWhitespace() {
        while (index < input.length && /\s/.test(input[index])) {
            index++;
        }
    }

    function consumeChar(char: string): boolean {
        consumeWhitespace();
        if (index < input.length && input[index] === char) {
            index++;
            return true;
        }
        return false;
    }

    function parseValue(): boolean {
        consumeWhitespace();

        if (input[index] === '"') {
            return parseString();
        } else if (input[index] === 't' || input[index] === 'f') {
            return parseBoolean();
        } else if (input[index] === 'n') {
            return parseNull();
        } else if (/[-0-9]/.test(input[index])) {
            return parseNumber();
        } else if (input[index] === '{') {
            return parseObject();
        } else if (input[index] === '[') {
            return parseArray();
        }

        return false;
    }

    function parseBoolean(): boolean {
        if (input.slice(index, index + 4) === 'true') {
            index += 4;
            return true;
        } else if (input.slice(index, index + 5) === 'false') {
            index += 5;
            return true;
        }
        return false;
    }

    function parseNull(): boolean {
        if (input.slice(index, index + 4) === 'null') {
            index += 4;
            return true;
        }
        return false;
    }

    function parseNumber(): boolean {
        const start = index;
        if (input[index] === '-') {
            index++;
        }
        while (index < input.length && /[0-9]/.test(input[index])) {
            index++;
        }
        if (input[index] === '.') {
            index++;
            while (index < input.length && /[0-9]/.test(input[index])) {
                index++;
            }
        }
        return index > start;
    }

    function parseString(): boolean {
        if (!consumeChar('"')) {
            return false;
        }
        while (index < input.length && input[index] !== '"') {
            if (input[index] === '\\') {
                index++;
            }
            index++;
        }
        if (!consumeChar('"')) {
            return false;
        }
        return true;
    }

    function parseArray(): boolean {
        if (!consumeChar('[')) {
            return false;
        }

        consumeWhitespace();

        if (consumeChar(']')) {
            return true;
        }

        while (true) {
            if (!parseValue()) {
                return false;
            }

            consumeWhitespace();

            if (!consumeChar(',')) {
                break;
            }

            consumeWhitespace();
        }

        if (!consumeChar(']')) {
            return false;
        }

        return true;
    }

    function parseKeyValuePair(): boolean {
        if (!parseString()) {
            return false;
        }

        consumeWhitespace();

        if (!consumeChar(':')) {
            return false;
        }

        consumeWhitespace();

        if (!parseValue()) {
            return false;
        }

        return true;
    }

    function parseObject(): boolean {
        if (!consumeChar('{')) {
            return false;
        }

        consumeWhitespace();

        if (consumeChar('}')) {
            return true;
        }

        while (true) {
            if (!parseKeyValuePair()) {
                return false;
            }

            consumeWhitespace();

            if (!consumeChar(',')) {
                break;
            }

            consumeWhitespace();
        }

        if (!consumeChar('}')) {
            return false;
        }

        return true;
    }

    return parseObject();
}
